### PR TITLE
Feature signup integration

### PR DESCRIPTION
**Pull request by ema999**
## Summary

In this branch I have added the functionality for the signup form.

- API call

- Service layer

- Show a loading bar

- Show errors

## Screenshots

Image-2020-11-04-at-1 06 05-PM

## Trello Card

https://trello.com/c/n4gwc0Jx

### CODE REVIEW

File: src/app/hooks/useRequest.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  useEffect(() => {
+    if (!payload || !Object.entries(payload).length) {
+      return;
+    }
+
+    sendRequest();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [payload, request]);
+  return [isLoading, submitError, sendRequest];
```
Comment (DavidRojas1612) [inline src/app/hooks/useRequest.js]: why not return the response?

File: src/config/api.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  }
+});
+
+api.addRequestTransform(request => {
+  if (request.data) {
+    request.data = deserializer.serialize(request.data);
+  }
+});
+
+export default api;
```
Comment (DavidRojas1612) [inline src/config/api.js]: add EOF

File: src/app/hooks/useRequest.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  const [submitError, setSubmitError] = useState([]);
+  const [response, setResponse] = useState();
+
+  const sendRequest = () => {
+    setIsLoading(true);
+    request(payload).then(data => {
+      if (data && data.ok) {
+        setSubmitError([]);
+        // eslint-disable-next-line no-console
+        console.log(data);
```
Comment (alejobh) [inline src/app/hooks/useRequest.js]: 👀

File: src/app/hooks/useRequest.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  useEffect(() => {
+    if (!payload || !Object.entries(payload).length) {
+      return;
+    }
+
+    sendRequest();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [payload, request]);
+  return [isLoading, submitError, sendRequest, response];
```
Comment (alejobh) [inline src/app/hooks/useRequest.js]: i would change the order, first loading, then error, then response and last the sendRequest

File: src/components/SignupForm/index.js [hunk 5]
```diff
@@ -1,79 +1,93 @@
...
             }
           })}
           error={errors[AUTH_INPUTS.confirmPassword]?.message}
         />
-        <ButtonForm type="submit" isFilled>
-          Sign Up
-        </ButtonForm>
-        <ButtonForm isDivider>Login</ButtonForm>
+        {isLoading ? (
+          <div>{t('FormValidations:loading')}...</div>
```
Comment (alejobh) [inline src/components/SignupForm/index.js]: this should be a `h3...h6` or `p`

File: src/components/SignupForm/index.js [hunk 6]
```diff
@@ -1,79 +1,93 @@
...
-    const user = {
+    setUserData({
       firstName,
       lastName,
       email,
       password,
       confirmPassword
-    };
-    console.log({ user });
+    });
```
Comment (alejobh) [inline src/components/SignupForm/index.js:26]: when do you fire the request?

File: src/config/api.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+api.addResponseTransform(response => {
+  if (response.data) {
+    response.data = serializer.serialize(response.data);
+  }
+});
+
+api.addRequestTransform(request => {
+  if (request.data) {
+    request.data = deserializer.serialize(request.data);
+  }
```
Comment (alejobh) [inline src/config/api.js:24]: this is for request, and i think its different here, i might be wrong

File: src/utils/errorHandler.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { t } from 'i18next';
+
+export const errorApiHandler = response => {
+  let error = t('Errors:apiError');
+
+  const { data } = response;
```
Comment (alejobh) [inline src/utils/errorHandler.js]: you dont have `data` if the network fails (`NETWORK_ERROR`) by default or there is a `CLIENT_ERROR`

File: src/app/hooks/useRequest.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+const useRequest = ({ request, payload }) => {
+  const [isLoading, setIsLoading] = useState();
+  const [submitError, setSubmitError] = useState([]);
+  const [response, setResponse] = useState();
+
+  const sendRequest = () => {
+    setIsLoading(true);
+    request(payload).then(data => {
+      if (data && data.ok) {
+        setSubmitError([]);
```
Comment (alejobh) [inline src/app/hooks/useRequest.js]: i think this empty array is wrong, should be null?

File: src/components/SignupForm/index.js (hunk 6 above)
Comment (ema999) [reply src/components/SignupForm/index.js:26]: the request is fired when we change the value of payload. If is necesary call again we can use sendRequest(payload)

File: src/config/api.js (hunk 7 above)
Comment (ema999) [reply src/config/api.js:24]: mm, Im using two diferents objects const serializer = new CamelcaseSerializer();
const deserializer = new SnakecaseSerializer(); Is working for request and response

File: src/components/SignupForm/index.js (hunk 6 above)
Comment (DavidRojas1612) [reply src/components/SignupForm/index.js:26]: Your useRequest hook should be able to give the component responsibility for when it wants to execute it

File: src/config/api.js (hunk 7 above)
Comment (DavidRojas1612) [reply src/config/api.js:24]: I see this well, you are using CamelcaseSerializer for the response, and SnakecaseSerializer for the request

Comment (alejobh) [reply src/config/api.js:24]: im refering to the `request` not about the serializer, but if it works, its ok

**Review by alejobh — APPROVED** [review]

**Review by DavidRojas1612 — APPROVED** [review]

**Pull request merged**